### DISCUSSION TITLE
$and mixed with $not conditions

### DISCUSSION THREAD

**Discussion by yackinn** [mikro-orm/mikro-orm#2962] [Q&A] [answered]
What's the correct way to achieve this behavior?
The conditions are two folded.

The conditions 1 and 2 should be met.
Also conditions 3 and 4 aren't allowed and should be ruled out.

```ts
$and: [
          {
            heldBy: user.id //1
          },
          {
            price: { $lt: 500 } //2
          },
          {
            $not: [
              {
                capacity: { $in: [0, 1] }  // 3
              },
              {
                price: 2    // 4
              }
            ]
          }
        ],
```

**Comment by B4nan** [comment, marked as answer]
`$not` is unary operator, so wrap the inner array into an object with `$and` or `$or`, depending on what behaviour you want.

**Reply by yackinn** [reply]
Perfect. Thank you.

In addition to that.

How do you add an or case to that keeping the $not applied to all cases?

```ts
$and: [
			// that's one combined positive match
          {
            heldBy: user['id']
          },
          {
            price: { $lt: 1000 }
          },

			// OR another positive match
		{
			amount: {$gt: 20}
		}

		// keeping $not applied to both of the above
          {
            $not: {
              $and: [
                {
                  capacity: { $in: [0, 1] }
                },
                {
                  price: 2
                }
              ]
            }
          }
        ],

```

**Reply by B4nan** [reply]
You need to think about it as the SQL query you want to build. `$not` is the same as wrapping expression in parens and prepending `NOT `, so `$not: { name: 'a' }` translates to `NOT (name = 'a')`, and `$not: { $or: [{ name: 'a' }, { name: 'b' }] }` translates to `NOT (name = 'a' or name = 'b')`.
